### ISSUE TITLE
[PFPurchase addObserverForProduct:] callback not called

### ISSUE THREAD

**Issue by iOSassou** [ParsePlatform/Parse-SDK-iOS-OSX#890]
Hello,

I have an app that is using v1.2.19 of the SDK and I want to update it to v1.12.0 to be able to migrate to a self-hosted Parse Server.

Everything works fine, except for In-App Purchases (consumable and non-consumable). [PFPurchase buyProduct:] correctly prompts the user to enter his Apple ID credential, ask to confirm the purchase and I receive the "Your purchase was successful" alert, but the associated block callback on [PFPurchase addObserverForProduct] is never called when the transaction succeed. Same thing for restoring an In-App Purchase.

If I revert back to the old version of the SDK, the block is correctly called so I know it's not a sandbox account issue or anything related to Apple servers.

I have the following log when the device is prompted to enter the credential and when I confirm the purchase:

```
Apr  6 16:39:14 iPad-Air akd[118] <Warning>: success
Apr  6 16:39:15 iPad-Air akd[118] <Warning>: success
Apr  6 16:39:15 iPad-Air accountsd[92] <Warning>: AIDA Notification plugin running
Apr  6 16:39:15 iPad-Air akd[118] <Error>: libMobileGestalt MGBasebandSupport.c:60: _CTServerConnectionCopyMobileEquipmentInfo: CommCenter error: 1:45 (Operation not supported)
Apr  6 16:39:15 iPad-Air akd[118] <Error>: libMobileGestalt MGBasebandSupport.c:189: No CT mobile equipment info dictionary while fetching kCTMobileEquipmentInfoIMEI
Apr  6 16:39:15 iPad-Air akd[118] <Error>: libMobileGestalt MGBasebandSupport.c:60: _CTServerConnectionCopyMobileEquipmentInfo: CommCenter error: 1:45 (Operation not supported)
Apr  6 16:39:15 iPad-Air akd[118] <Error>: libMobileGestalt MGBasebandSupport.c:189: No CT mobile equipment info dictionary while fetching kCTMobileEquipmentInfoMEID
Apr  6 16:39:15 iPad-Air akd[118] <Error>: libMobileGestalt MGBasebandSupport.c:60: _CTServerConnectionCopyMobileEquipmentInfo: CommCenter error: 1:45 (Operation not supported)
Apr  6 16:39:15 iPad-Air akd[118] <Error>: libMobileGestalt MGBasebandSupport.c:189: No CT mobile equipment info dictionary while fetching kCTMobileEquipmentInfoICCID
Apr  6 16:39:15 iPad-Air MobileGestaltHelper[776] <Error>: libMobileGestalt MobileGestalt.c:281: server_access_check denied access to question UniqueDeviceID for pid 862
Apr  6 16:39:15 iPad-Air syncdefaultsd[862] <Error>: libMobileGestalt MobileGestaltSupport.m:151: pid 862 (syncdefaultsd) does not have sandbox access for re6Zb+zwFKJNlkQTUeT+/w and IS NOT appropriately entitled
Apr  6 16:39:15 iPad-Air syncdefaultsd[862] <Error>: libMobileGestalt MobileGestalt.c:542: no access to UniqueDeviceID (see <rdar://problem/11744455>)
Apr  6 16:39:16 iPad-Air itunesstored[98] <Warning>: Could not load library [21]
Apr  6 16:39:16 iPad-Air itunesstored[98] <Warning>: Could not load library [21]
Apr  6 16:39:16 iPad-Air itunescloudd[472] <Warning>: Updating media purchase history for initial load...
Apr  6 16:39:16 iPad-Air limitadtrackingd[345] <Warning>: Current store account has changed from (null) to 489711460.
Apr  6 16:39:16 iPad-Air itunescloudd[472] <Warning>: Updating media purchase history for initial load...
Apr  6 16:39:16 iPad-Air itunescloudd[472] <Warning>: Updating media purchase history for initial load...
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  clearing all local changes that had been scheduled for push
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  reseting sync anchor to 0, and scheduling pull from server
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  target sync date from client: 2016-04-06 14:39:22 +0000 (in 4.99 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  setting target date to: 2016-04-06 14:39:22 +0000 (in 62632257642.72 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  scheduling sync (via BackgroundTaskJob) 4.990217 seconds from now...
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  clearing all local changes that had been scheduled for push
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  reseting sync anchor to 0, and scheduling pull from server
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  target sync date from database: 2016-04-06 14:39:22 +0000 (in 4.95 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  scheduling sync (via BackgroundTaskJob) 4.946218 seconds from now...
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  clearing all local changes that had been scheduled for push
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  reseting sync anchor to 0, and scheduling pull from server
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  target sync date from database: 2016-04-06 14:39:22 +0000 (in 4.91 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  scheduling sync (via BackgroundTaskJob) 4.912434 seconds from now...
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  clearing all local changes that had been scheduled for push
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  reseting sync anchor to 0, and scheduling pull from server
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  target sync date from database: 2016-04-06 14:39:22 +0000 (in 4.89 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  scheduling sync (via BackgroundTaskJob) 4.889297 seconds from now...
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  clearing all local changes that had been scheduled for push
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  reseting sync anchor to 0, and scheduling pull from server
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  target sync date from database: 2016-04-06 14:39:22 +0000 (in 4.85 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  scheduling sync (via BackgroundTaskJob) 4.850278 seconds from now...
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  clearing all local changes that had been scheduled for push
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  reseting sync anchor to 0, and scheduling pull from server
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  target sync date from database: 2016-04-06 14:39:22 +0000 (in 4.70 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  scheduling sync (via BackgroundTaskJob) 4.702794 seconds from now...
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  clearing all local changes that had been scheduled for push
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  reseting sync anchor to 0, and scheduling pull from server
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  target sync date from database: 2016-04-06 14:39:22 +0000 (in 4.69 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  scheduling sync (via BackgroundTaskJob) 4.692428 seconds from now...
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  clearing all local changes that had been scheduled for push
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  reseting sync anchor to 0, and scheduling pull from server
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  target sync date from database: 2016-04-06 14:39:22 +0000 (in 4.65 sec)
Apr  6 16:39:17 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  scheduling sync (via BackgroundTaskJob) 4.652937 seconds from now...
Apr  6 16:39:22 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDDomainSyncedStorageController]  running synchronizeImmediatelyWithCompletionHandler: now
Apr  6 16:39:22 iPad-Air storebookkeeperd[128] <Warning>: [StoreBookkeeper] -[SBKUniversalPlaybackPositionStore _onQueueRunTaskWithName:taskCompletionHandler:runTaskBlock:] _onQueueSync - beginning synchronize operation...
Apr  6 16:39:22 iPad-Air storebookkeeperd[128] <Warning>: [StoreBookkeeper] -[SBKUniversalPlaybackPositionStore _onQueueLoadBagContextWithCompletionHandler:] _onQueueLoadBag - load bag operation beginning...
Apr  6 16:39:22 iPad-Air storebookkeeperd[128] <Warning>: [StoreBookkeeper] __81-[SBKUniversalPlaybackPositionStore _onQueueLoadBagContextWithCompletionHandler:]_block_invoke413 _onQueueLoadBag - load bag completed.
Apr  6 16:39:22 iPad-Air storebookkeeperd[128] <Warning>: [StoreBookkeeper] __89-[SBKUniversalPlaybackPositionStore _onQueueSynchronizeImmediatelyWithCompletionHandler:]_block_invoke _onQueueSync - synchronize operation beginning...
Apr  6 16:39:23 iPad-Air storebookkeeperd[128] <Warning>: [StoreBookkeeper] [SBKTransactionController.m:447] Transaction Failed: <SBKSyncTransaction: 0x134db2520> domain = com.apple.upp, SYNC: anchor=0 -- Error: <SBKStoreError:0x136077180> "Error Code = SBKStoreErrorCodeStoreAccountSessionExpired (-1004)"
Apr  6 16:39:23 iPad-Air storebookkeeperd[128] <Warning>: [StoreBookkeeper] __96-[SBKUniversalPlaybackPositionStore _onQueueRunTaskWithName:taskCompletionHandler:runTaskBlock:]_block_invoke281 _onQueueSync - synchronize completed.  ERROR = 'Error Code = SBKStoreErrorCodeStoreAccountSessionExpired (-1004)'.
Apr  6 16:39:23 iPad-Air storebookkeeperd[128] <Warning>: [UPP-SBDJobScheduler]  Could not synchronize domain 'com.apple.upp' (<SBKStoreError:0x13607d850> "Error Code = SBKStoreErrorCodeStoreAccountSessionExpired (-1004)").  Synchronization will be re-attempted when the network connectivity, or account status has changed.
```

**Comment by nlutsenko** [comment]
Hey @iOSassou, thanks for reporting, let's debug this together and find the reason behind it:
- Is the block for `PFPurchase.buyProduct()` called or the same happens here - it's not called in the same way the observer for a product is not?
- Did you try using the latest SDK with the same server your old SDK is using or only with open source Parse Server?
- Any chance you can instrument your app with our Network Debug Tool from [this guide](https://github.com/ParsePlatform/Parse-SDK-iOS-OSX/wiki/Network-Debug-Tool) and console log for network requests that happen when you are making the purchase here?

**Comment by iOSassou** [comment]
- [PFPurchase buyProduct] block is called, I use it for error checking and my error message is displayed if I cancel the transaction for example.
- I'm not using open source Parse Server. I'm updating to the latest SDK to prepare the migration, but I'm still hitting still api.parse.com
- No problem, I will be able to instrument the app tomorrow if it's needed

**Comment by nlutsenko** [comment]
Sounds good - looking forward to see the logs here.

**Comment by iOSassou** [comment]
I followed the guide, but nothing gets logged during the purchase. I tested with other parts of my app and I can see the Parse URL/Methods/Header/etc. in my console, but nothing logged during the Apple authentification and purchase of a product.

I've also added extra NSLogs and I can confirm that the [PFPurchase buyProduct] block is called, but not the one for [PFPurchase addObserverForProduct].

Deployment target is iOS 8.0 and I'm using the latest version of Xcode (7.3) with iOS 9.3 Base SDK.

**Comment by nlutsenko** [comment]
@iOSassou, very strange, since I've just tested on iOS 9.3 and it works just fine.
Any chance you can trace `PFPaymentTransactionObserver` class on a `PFPurchaseController`?
The part that I am explicitly looking for is whether it has registered blocks when `addObserverForProduct:block:` is getting registered.

Also, few more questions - are you setting up the observer for the product before calling into `buyProduct` or after? If after - there is a chance that it registers in a wrong way.

**Comment by iOSassou** [comment]
I set up all my observers in the AppDelegate, so it's very early in the app lifecycle. However, after your comment I tried to change the order of my setup and I was able to make PFPurchase work.

In v1.12.0, you must set up `PFPurchase` observers after `[Parse setApplicationId:XXX clientKey:XXX]`. It was not the case in v1.2.19 so some changes made the two classes dependent on each other.

Thank you for your help
